File: src/client/python/src/grant.py
from src.network.network import get, post
from sessionless import SessionlessSecp256k1
import time

async def grant(uuid, destinationUUID, amount, description, get_keys):
    sessionless = SessionlessSecp256k1()

    timestamp = f'{round(time.time() * 1000)}'
    message = f'{timestamp}{uuid}{destinationUUID}{amount}{description}'
    print(message)
    signature = await sessionless.sign(message.encode('ascii'), get_keys)

    payload = {
        "timestamp": timestamp,
        "uuid": uuid,
        "destinationUUID": destinationUUID,
        "amount": amount,
        "description": description,
        "signature": signature
    }

    response = await post(f'user/{uuid}/grant', payload)

    return response

File: src/server/node/src/routes/grant.js
import user from '../user/user.js';
import experience from '../experience/experience.js';
import sessionless from 'sessionless-node';

const grant = async (req, res) => {
  const body = req.body;
  const timestamp = body.timestamp;
  const uuid = req.params.uuid;
  const destinationUUID = body.destinationUUID;
  const amount = body.amount;
  const description = body.description;
  const signature = body.signature;

  const message = timestamp + uuid + destinationUUID + amount + description;
  const sourceUser = await user.getUser(uuid);

  if(!sessionless.verifySignature(signature, message, sourceUser.pubKey)) {
console.log('auth erere');
console.log(message);
    res.status(403);
    return res.send({error: 'Auth error'});
  } 
console.log('here');
  const destinationUser = await user.getUser(destinationUUID);
console.log('there');
  const updatedUser = await experience.grant(sourceUser, destinationUser, amount);

console.log('sending back', updatedUser);

  res.send(updatedUser);
};

export { grant };
